sql/src/main/java/io/crate/analyze/MultiSourceSelect.java

```diff
@@ -23,56 +23,107 @@
 
 import io.crate.analyze.relations.*;
 import io.crate.analyze.symbol.Field;
+import io.crate.analyze.symbol.FieldReplacer;
 import io.crate.analyze.symbol.Symbol;
+import io.crate.collections.Sets2;
+import io.crate.metadata.Functions;
 import io.crate.metadata.Path;
+import io.crate.metadata.TransactionContext;
 import io.crate.metadata.table.Operation;
 import io.crate.sql.tree.QualifiedName;
 
 import javax.annotation.Nonnull;
 import java.util.*;
-import java.util.stream.Collectors;
+import java.util.function.Function;
 
 public class MultiSourceSelect implements QueriedRelation {
 
-    private final Map<QualifiedName, RelationSource> sources;
-    private final QuerySpec querySpec;
+    private final Map<QualifiedName, AnalyzedRelation> sources;
     private final Fields fields;
     private final List<JoinPair> joinPairs;
     private final Set<Symbol> requiredForQuery;
     private final Set<Field> canBeFetched;
     private final Optional<RemainingOrderBy> remainingOrderBy;
     private QualifiedName qualifiedName;
-
-    public static MultiSourceSelect createWithPushDown(MultiSourceSelect mss, QuerySpec querySpec) {
+    private QuerySpec querySpec;
+
+    /**
+     * Create a new MultiSourceSelect which has it's sources transformed into subQueries which contain
+     * any expressions that only affect a single source.
+     * Example:
+     * <pre>
+     *     select t1.x, t2.y from t1, t2 where t1.x = 10
+     *
+     *     becomes
+     *
+     *     select t1.x, t2.y from
+     *          (select x from t1 where x = 10) t1,
+     *          (select y from t2)
+     * </pre>
+     */
+    public static MultiSourceSelect createWithPushDown(Functions functions,
```

Comment:
Think we could consider having a new MultiSourceSelect class that has `sources <QualifiedName, QueriedRelation>` to avoid all the casts that became otherwise necessary. Thoughts?